src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs

```diff
@@ -1938,8 +2007,85 @@ private void AddMeshData(string id, HelixRenderPackage rp, int index, int count,
 
             meshGeometry3D.Geometry = mesh;
             meshGeometry3D.Name = baseId;
+            meshGeometry3D.Tag = id;
+
+            if (intances != null)
+            {
+                meshGeometry3D.Instances = intances;
+            }
+        }
+
+        private void AddLineData(string id, HelixRenderPackage rp, int index, int count, bool drawDead, string baseId, List<Matrix> intances = null)
+        {
+            FastList<Vector3> lPositions;
+            FastList<Color4> lColors;
+            FastList<int> lIndices;
+
+            var l = rp.Lines;
+            if (index == 0 && count == rp.Lines.Positions.Count)
+            {
+                lPositions = l.Positions;
+                lColors = l.Colors;
+                lIndices = l.Indices;
+            }
+            else
+            {
+                lPositions = l.Positions.GetRange(index, count);
+                lColors = l.Colors.GetRange(index, count);
+                var firstIndicesIndex = l.Indices.IndexOf(index);
+                var indicesCount = l.Indices.IndexOf(index + count - 1) - firstIndicesIndex + 1;
+                lIndices = l.Indices.GetRange(firstIndicesIndex, indicesCount);
+            }
+
+            LineGeometryModel3D lineGeometry3D;
+
+            if (Element3DDictionary.ContainsKey(id))
+            {
+                lineGeometry3D = Element3DDictionary[id] as LineGeometryModel3D;
+            }
+            else
+            {
+                // If the package contains mesh vertices, then the lines represent the 
+                // edges of meshes. Draw them with a different thickness.
+                lineGeometry3D = CreateLineGeometryModel3D(rp, rp.MeshVertices.Any() ? 0.5 : 1.0);
+                Element3DDictionary.Add(id, lineGeometry3D);
+            }
+
+            var lineSet = lineGeometry3D.Geometry == null ? HelixRenderPackage.InitLineGeometry()
+                        : lineGeometry3D.Geometry as LineGeometry3D;
+            var previousPositionCount = lineSet.Positions.Count;
+
+
+            lineSet.Positions.AddRange(lPositions);
+
+
+            if (drawDead)
+            {
+                lineSet.Colors.AddRange(Enumerable.Repeat(defaultDeadColor, lPositions.Count));
+            }
+            else
+            {
+                lineSet.Colors.AddRange(lColors.Any()
+                 ? lColors
+                 : Enumerable.Repeat(defaultLineColor, lColors.Count));
+            }
+
+            //colorCache[id] = lineSet.Colors;
+
+            var adjustment = previousPositionCount - lIndices[0];
+
+            lineSet.Indices.AddRange(lIndices.Select(i => i + adjustment));
+
+            lineGeometry3D.Geometry = lineSet;
+            lineGeometry3D.Name = baseId;
+
+            if (intances != null)
+            {
+                lineGeometry3D.Instances = intances;
+            }
         }
```

Comment:
Can we add more comments to this function to explain what's going on? It seems very logic-heavy.